docs/source/developer/concepts.rst

```diff
@@ -0,0 +1,307 @@
+    template <class L>
+    const_iterator cbegin() const noexcept;
+    template <class L>
+    const_iterator cend() const noexcept;
+
+    template <class L>
+    const_reverse_iterator rbegin() const noexcept;
+    template <class L>
+    const_reverse_iterator rend() const noexcept;
+    template <class L>
+    const_reverse_iterator crbegin() const noexcept;
+    template <class L>
+    const_reverse_iterator crend() const noexcept;
+
+.. image:: iteration.svg
+
+This template parameter is defaulted to ``DEFAULT_LAYOUT`` (see :ref:`configuration-label`), so
+that `xtensor` expressions can be used in generic code such as:
+
+.. code::
+
+    std::copy(a.cbegin(), a.cend(), b.begin());
+
+where ``a`` and ``b`` can be arbitrary types (from `xtensor`, the STL or any external library)
+supporting standard iteration.
+
+``xiterable`` inherits from ``xconst_iterable`` and provides non-const counterpart of methods
+defined in ``xconst_iterable``. Like ``xexpression``, both are CRTP classes whose template
+parameter must be the most derived type.
+
+Besides traditional methods for iterating, ``xconst_iterable`` and ``xiterable`` provide overloads
+taking a shape parameter. This allows to iterate over an expression as if it was broadcast to the
+given shape:
+
+.. code::
+
+    #include <algorithm>
+    #include <iterator>
+    #include <iostream>
+    #include "xtensor/xarray.hpp"
+
+    int main(int argc, char* argv[])
+    {
+        xt::xarray<int> a = { 1, 2, 3 };
+        std::vector<std::size_t> shape = { 2, 3 };
+        std::copy(a.cbegin(shape), a.cend(shape), std::output_iterator(std::cout, " "));
+        // output: 1 2 3 1 2 3
+    }
+
+Iterators returned by methods defined in ``xconst_iterable`` and ``xiterable`` are random access
+iterators.
+
+.. _xsemantic-concept-label:
+
+xsemantic
+~~~~~~~~~
+
+The ``xsemantic_base`` interface provides methods for assigning an expression:
+
+.. code::
+
+    template <class E>
+    disable_xexpression<E, derived_type&> operator+=(const E&);
+
+    template <class E>
+    derived_type& operator+=(const xexpression<E>&);
+
+and similar methods for ``operator-=``, ``operator*=``, ``operator/=``, ``operator%=``,
+``operator&=``, ``operator|=`` and ``operator^=``.
+
+The first overload is meant for computed assignment involving a scalar; it allows to write code like
+
+.. code::
+
+    #include "xtensor/xarray.hpp"
+    #include "xio.hpp"
+
+    int main(int argc, char* argv)
+    {
+        xarray<int> a = { 1, 2, 3 };
+        a += 4;
+        std::cout << a << std::endl;
+        // outputs { 5, 6, 7 }
+    }
+
+We rely on SFINAE to remove this overload from the overload resolution set when the parameter that we want
+to assign is not a scalar, avoiding ambiguity.
+
+All operator-based methods taking an ``xexpression`` parameter do not perform a direct assignment, instead
+they compute a temporary before calling the ``assign`` method. Thus, if ``a`` and ``b`` are expressions,
```

Comment:
Operator-based method taking a general `xexpression` parameter don't perform a direct assignment. Instead, the result is assigned to a temporary variable first, in order to prevent issues with aliasing.